src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -0,0 +1,2293 @@
+    symbol_table - symbol table to which the class will be added
+
+ Purpose: Add to the symbol table type declaration for a String-like
+          Java class.
+
+\*******************************************************************/
+
+void java_string_library_preprocesst::add_string_type(
+  const irep_idt &class_name, symbol_tablet &symbol_table)
+{
+  class_typet string_type;
+  string_type.set_tag(class_name);
+  string_type.components().resize(3);
+  string_type.components()[0].set_name("@java.lang.Object");
+  string_type.components()[0].set_pretty_name("@java.lang.Object");
+  string_type.components()[0].type()=symbol_typet("java::java.lang.Object");
+  string_type.components()[1].set_name("length");
+  string_type.components()[1].set_pretty_name("length");
+  string_type.components()[1].type()=java_int_type();
+  string_type.components()[2].set_name("data");
+  string_type.components()[2].set_pretty_name("data");
+  // Use a pointer-to-unbounded-array instead of a pointer-to-char.
+  // Saves some casting in the string refinement algorithm but may
+  // be unnecessary.
+  string_type.components()[2].type()=pointer_typet(
+    array_typet(java_char_type(), infinity_exprt(java_int_type())));
+  string_type.add_base(symbol_typet("java::java.lang.Object"));
+
+  symbolt string_symbol;
+  string_symbol.name="java::"+id2string(class_name);
+  string_symbol.base_name=id2string(class_name);
+  string_symbol.type=string_type;
+  string_symbol.is_type=true;
+
+  symbol_table.add(string_symbol);
+}
+
+/*******************************************************************\
+
+Function: fresh_array
+
+  Inputs:
+    type - an array type
+    location - a location in the program
+    symbol_table - symbol table
+
+  Output: a new symbol
+
+ Purpose: add a symbol in the table with static lifetime and name
+          containing `cprover_string_array` and given type
+
+\*******************************************************************/
+
+symbol_exprt java_string_library_preprocesst::fresh_array(
+  const typet &type,
+  const source_locationt &location,
+  symbol_tablet &symbol_table)
+{
+  symbolt array_symbol=get_fresh_aux_symbol(
+    type,
+    "cprover_string_array",
+    "cprover_string_array",
+    location,
+    ID_java,
+    symbol_table);
+  array_symbol.is_static_lifetime=true;
+  return array_symbol.symbol_expr();
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::declare_function
+
+  Inputs:
+    function_name - a name
+    type - a type
+    symbol_table - symbol table
+
+ Purpose: declare a function with the given name and type
+
+\*******************************************************************/
+
+void java_string_library_preprocesst::declare_function(
+  irep_idt function_name, const typet &type, symbol_tablet &symbol_table)
+{
+  auxiliary_symbolt func_symbol;
+  func_symbol.base_name=function_name;
```

Comment:
Similarly, set pretty_name; most code assumes pretty-name may not be defined and falls back to base_name but I've seen cases where this is not true.